### DISCUSSION TITLE
Hidden fields, computed fields, init-only fields?

### DISCUSSION THREAD

**Discussion by naught101** [pydantic/pydantic#3290] [Question]
I have a class like this:

```python
class Model(BaseModel):

    things: tuple = Field(
        ..., description="List of things to include"
    )

    num_things: int = None

    def __init__(self, **data):
        super().__init__(**data)

        self.num_things = len(self.things)
```

Which works, except that:

- I **don't** want `num_things` to be allowed as an input - I want it to be compute-only, based on whatever the length of `things` is.
- I need to provided a default for it, or pydantic complains about a missing value. Providing an `int` default seems unsensible when I don't know what it should be.
- if I provide `None` as the default in the field def, as above, mypy complains.
- I **don't** want it to be returned using the `.dict()` method (since it's strictly computable from `things` at all times).
- I also **don't** want it to appear in the class `.schema()`.
- I just want it to be accessible in code, in order to save a bit of computation.

Is there a way to do this? It seems like this is what `dataclasses.InitVar` is for, but I can't figure out if there's a `BaseModel` equivalent.

**Comment by PrettyWood** [comment]
Hi @naught101 
The best is probably to wait for computed fields in https://github.com/samuelcolvin/pydantic/pull/2625, which will be released in v1.9
```py
from pydantic import BaseModel, Field, computed_field

class Model(BaseModel):

    things: tuple = Field(
        ..., description="List of things to include"
    )

    @computed_field
    def num_things(self) -> int:
        return len(self.things)


m = Model(things=(1, 2, 3))
assert m.num_things == 3
```

**Comment by MikesRND** [comment]
Would a standard python property suffice?  e.g.
```python
from pydantic import BaseModel, Field

class Model(BaseModel):

    things: tuple = Field(
        ..., description="List of things to include"
    )

    @property
    def num_things(self) -> int:
        return len(self.things)


m = Model(things=(1, 2, 3))
assert m.num_things == 3
print(m)
print(m.dict())
```
returns
```
things=(1, 2, 3)
{'things': (1, 2, 3)}
```

**Reply by naught101** [reply]
Yes, this makes sense, and works how I want it. In some cases I would prefer not to do the computation every time, but in most cases this would be fine, and it's not hard to put in a simple cache that stores the value the first time it's calculated. Thanks!

**Reply by naught101** [reply]
Ah, I can just use `@cached_property` for that :blush:

**Reply by naught101** [reply]
In combination with `keep_untouched`, see https://github.com/samuelcolvin/pydantic/issues/1241

**Comment by martijnmathijsen** [comment]
I came across this item when looking for attributes/parameters only used during initialization of `pydantic.BaseModel`.

For `dataclass` there is the option to use `dataclass.InitVar`, however this is not supported for `pydantic.BaseModel`.

@naught101 as you mentioned a similar question here, have you found a way to use init-only fields for BaseModel?

**Reply by manmartgarc** [reply]
Following up on this, were you able to find a solution? I'm looking at exactly what you mention, i.e. `InitVar` from dataclasses

Edit: found the answer in `init_var=True`; see [docs](https://docs.pydantic.dev/latest/concepts/fields/#dataclass-constraints)
